proxy/coco/src/state.rs

```diff
@@ -103,40 +97,54 @@ where
     P: TryInto<PersonPayload> + Send,
     Error: From<P::Error>,
 {
-    match peer
-        .using_storage(move |store| local::default(store))
+    if let Some(owner) = default_owner(peer).await? {
+        return Ok(owner);
+    }
+
+    let payload = payload.try_into()?;
+    let pk = keys::PublicKey::from(peer.signer().public_key());
+    let delegations = Some(pk).into_iter().collect();
+    let person = peer
+        .using_storage(move |store| person::create(store, payload, delegations))
+        .await??;
+
+    let urn = person.urn();
+    let owner = peer
+        .using_storage(move |store| local::load(store, urn))
         .await??
-    {
-        Some(owner) => Ok(owner),
-        None => {
-            let pk = keys::PublicKey::from(peer.signer().public_key());
-            let payload = payload.try_into()?;
-            let person = peer
-                .using_storage(move |store| {
-                    person::create(store, payload, Some(pk).into_iter().collect())
-                })
-                .await??;
-
-            let urn = person.urn();
-            let owner = peer
-                .using_storage(move |store| local::load(store, urn))
-                .await??
-                .ok_or_else(|| Error::PersonNotFound(person.urn()))?;
-
-            {
-                let owner = owner.clone();
-                peer.using_storage(move |store| {
-                    let mut config = store.config()?;
-                    config.set_user(owner)?;
-
-                    Ok::<_, Error>(())
-                })
-                .await??;
-            }
+        .ok_or_else(|| Error::PersonNotFound(person.urn()))?;
 
-            Ok(owner)
-        },
-    }
+    set_default_owner(peer, owner.clone()).await?;
+
+    Ok(owner)
+}
+
+/// Sets a new person payload for the default owner of this [`Peer`].
+///
+/// # Errors
+///
+///   * Fails to load the default owner
+///   * Fails to verify `User`.
+///   * Fails to set the default `rad/self` for this `PeerApi`.
+#[allow(clippy::single_match_else)]
+pub async fn update_owner_payload<P>(
+    peer: &Peer<BoxedSigner>,
+    payload: P,
+) -> Result<LocalIdentity, Error>
+where
+    P: TryInto<PersonPayload> + Send,
+    Error: From<P::Error>,
+{
+    let urn = default_owner(peer).await?.ok_or(Error::MissingOwner)?.urn();
+    let payload = payload.try_into()?;
+    peer.using_storage(move |store| person::update(store, &urn, None, payload, None))
+        .await??;
+    // Update the default user info in the config with the new payload data.
+    // This works, because the payload update doesn't change the owner's URN,
+    // so the old config can be used to load the owner with the updated payload.
+    let owner = default_owner(peer).await?.ok_or(Error::MissingOwner)?;
+    set_default_owner(peer, owner.clone()).await?;
```

Comment:
I'm not sure this will always work in the future. If the `payload` is updating the document with a new key (or new keys), then the `Person` won't be verified since it won't necessarily have a quorum of signatures.

We're getting away with it now because there's only one key in the document and the signing of that key is happening in `update`, I think.

Reply:
Could you point me in a direction where this could be implemented to be future-proof regarding your point?

Reply:
So [these tests](https://github.com/radicle-dev/radicle-link/blob/master/librad/src/identities/git/tests/person.rs) give us an idea of what's going on here. The concept being that if I'm working on a Radicle from my desktop and my laptop I'll have two separate keys for them, but I'd like to use the one `Person` document for `fintohaps` alter-ego.

As soon as I [`update`](https://github.com/radicle-dev/radicle-link/blob/master/librad/src/git/identities/person.rs#L123-L148) my document, on my desktop, with my eth address to (GET PAID :money_mouth_face:) I'll have a document that is signed but not verified. This change needs to be propagated to my laptop and I [`merge`](https://github.com/radicle-dev/radicle-link/blob/master/librad/src/git/identities/person.rs#L150-L172) the changes if I'm happy that it was me. I think the desktop side will also have to `merge` for _it_ to be in a verified state.

I think but I'm not sure that, running `set_default_owner` here _without_ verifying will mean that it fails because `LocalIdentity` is a wrapper around `VerifiedPerson`. If it's the case that it doesn't fail my bet is that it won't have the update you think it'll have and instead it will be at the previously verified state (sans eth address) -- [this comment](https://github.com/radicle-dev/radicle-link/blob/master/librad/src/git/identities/person.rs#L57-L62) alludes to that.

So this updating of documents across devices needs be incorporated into the plans, in my opinion. I've mentioned it several times that we need to work out this flow if we want to open up the new features `radicle-link` now provides. It'll be the same for adding a maintainer to a `Project` :grin: I'm unsure of what thoughts @radicle-dev/product has had around this flow of cross-device verification, but might be worth thinking about as well.

As for this specific function, I would think that we would need to figure out when a document needs verification from other keys (in the case of `Person`) -- I'm not sure `radicle-link` provides this, but maybe @kim can chime in on that point.

Reply:
> As for this specific function, I would think that we would need to figure out when a document needs verification from other keys (in the case of Person) -- I'm not sure radicle-link provides this, but maybe @kim can chime in on that point.

The identity history permits commits near the tip of the `rad/id` branch which don't pass the quorum rules. This means that, if the oid the tip is pointing to is not equal to the `content_id` of the verified identity, some action is required. This can be determined using `get` vs. `verify` (we could also add some sugar to call `verify` on the result of `get`, but haven't).

Afaics, updating the "default owner" doesn't make much sense here indeed -- the default it just the URN stored in the git config, and since that is read at the beginning, it is presumably still the same. That is, `person::update` is enough for what you want to do here, I think.

Reply:
I'm thinking it would make sense to carry the information in the result of `verify`. At the point we know that it's not current tip so it could be something like: `verify -> Urn -> Verifying<Person>`, where:
```rust
enum Verifying<A> {
  Latest(A),
  Behind(A),
}
```

Reply:
That doesn’t seem like a huge improvement over a Boolean. If you’d want to rework the fold so it keeps accumulating non-quorum updates after it found a quorum one (similar to how the validation applicative works), I would certainly meet that with appreciation.

Reply:
> Afaics, updating the "default owner" doesn't make much sense here indeed -- the default it just the URN stored in the git config, and since that is read at the beginning, it is presumably still the same.

The update of the default owner updates the name and the email of the git user too. It's not very useful for our short-term use case (update of the Ethereum claim), but at some point it probably should be a part of the update of the metadata. I've added it only for the sake of completeness and to make the API behavior consistent. We can remove it and document that the the metadata update mustn't update the user name.

Reply:
> The update of the default owner updates the name and the email of the git user too

Eh? Where does it do that?

Reply:
It updates `user.name`, that’s correct. `user.email` is the peer id, so doesn’t change.

Both are only used for commits which are made by `librad`, and are not normally user-visible. They are set only because git expects _something_ in the committer/author fields, we never actually inspect them.

Reply:
> It's not very useful for our short-term use case [..] We can remove it and document that the the metadata update mustn't update the user name.

I'm not advocating we remove this :) I'm thinking about the _long-term_ of these updates since it would benefit _all_ cases of updating `Person` documents -- and also `Project` documents.

Reply:
IIUC:
1. The app doesn't support multi-device, quorum signatures and won't know what to do with the knowledge that an update isn't validated yet
2. Updating git's `user.name` and `user.email` is pointless, nobody will ever see it, especially in the app

Maybe let's stick to an MVP, keep this PR as it is but remove the `set_default_owner` part?

Reply:
Thank you for your reactions @xla and @FintanH, but could you elaborate? Do you have a better plan and what are the steps to implement it?

Reply:
Sure, was composing my thoughts. It's a bit disappointing to punt on helping the implementation of the protocol to both improve the protocol _and_ the upstream client. Is there a rush to get this in with the current semantics?

As I pointed out, it would be useful to bring to the surface when identities are out of date and need interaction. Kim pointed to some possible ways of looking at how to do this. A plan would be to evaluate these methods and implement them.

Reply:
Unfortunately yes, there is a rush to push this out in some form to unblock the rest of the team 🙁 The mechanism for coping with out of date identities looks like a good material for a separate PR. IMO this feature won't be deeply affected when the HTTP API will be extended with information about freshness of an identity (which will be useful only when we add logic to app to handle it and will probably affect multiple HTTP endpoints) or a mechanism for resolution is added to coco.

Reply:
Ja, alles klar. `coco` is on its way over to `radicle-link` now anyway, so it'll be easier to add these things in.